In [1]:
"""
Full Name : Kutay Eroğlu
Student ID : 59970
Assignment No/Name : 4 / Population Simulation
Version : 1.1

"""

'\nFull Name : Kutay Eroğlu\nStudent ID : 59970\nAssignment No/Name : 4 / Population Simulation\nVersion : 1.1\n\n'

In [2]:
import numpy as np 
from datetime import datetime as dt 

In [3]:
population_data = {'Males':0,'Females':0,'Couples':0,'Parents':0,'Baby girls':0,'Baby boys':0} 
simulation_status = 1 

In [4]:
def simulation(scenario, N , MFR, FR, NGEN) : 
    q = dt.now()   
    
    ####### Calculating basic population data #######
    population_data['Males'] = int(N * MFR)
    population_data['Females'] = N - population_data['Males']
    
    print(scenario, " Scenario:\n- - - - - - - - - -")
    
    for num in range (0, NGEN) :         
        population_data['Couples'] = min(population_data['Males'], population_data['Females'])
        population_data['Parents'] = int(population_data['Couples'] * FR )
        population_data['Baby girls'] = population_data['Baby boys'] = 0

        if(scenario == 'One Child'):
            ####### Zeros in the children_arr correspond to girl, ones to boy respectively #######
            children_arr = np.random.randint(2,size = population_data['Parents'])    
            for i in children_arr.tolist() : 
                if (i == 0) : 
                    population_data['Baby girls'] += 1
                else : 
                    population_data['Baby boys'] += 1

        elif(scenario == 'One Son'):
            children_arr = np.random.randint(2, size = population_data['Parents']*5)
            counter = 0 
            
            for child in children_arr : 
                if(len(children_arr) - 4 <= counter) : 
                    break
                ## If this condition holds true, then all five children are female 
                if(1 not in children_arr[counter: counter+5]) : 
                    population_data['Baby girls'] += 5
                    counter += 5
                else : 
                    for i in range (0,5) :
                        if(children_arr[counter + i] == 0) : 
                            population_data['Baby girls'] += 1 
                        elif(children_arr[counter + i] == 1) : 
                            population_data['Baby boys'] += 1
                            counter += (5-i+1) 
                            break
                        
        else :
            raise ValueError("Unavailable scenario.")
        
        print("Generation ", num+1 ," Statistics: ")
        print("No. of MALES : ", population_data['Males'])
        print("No. of FEMALES: ", population_data['Females'])
        print("No. of couples : ", population_data['Couples'])
        print("No. of Parents : ", population_data['Parents'])
        print("No. of babies Total : ", population_data['Baby girls'] + population_data['Baby boys'])
        print("No. of baby girls : ", population_data['Baby girls'])
        print("No. of baby boys : ", population_data['Baby boys'], "\n")   
        
        ####### Entering new generation, parent data should be updated #######        
        population_data['Males'] = population_data['Baby boys']
        population_data['Females'] = population_data['Baby girls']
        
    print('--------- SURE - ',dt.now() - q, "\n")    

## If the user wants to end the simulation, they should enter a negative value for any of the inputs or they can provide the indicated input when they are asked whether they want to carry out another simulation.

In [5]:
while(simulation_status) : 
    
    ########Initial Population#######
    N = int(input("Enter initial population: "))
    
    if(N <= 1) :
        raise ValueError("Initial population can't be smaller than two due to simulation purposes.")
    
    #######Male/Female ratio (to be used for new born babies’ gender)####### 
    MFR = float(input("Enter Male/Female ratio: "))
    
    if(MFR <= 0) : 
        raise ValueError("Initial Male/Female ratio can't be smaller than or equal to zero due to simulation purposes.")
    
    #######Fertility rate (the % of parents that can have a baby)#######
    FR = float(input("Enter fertility rate with as fraction value (e.g 0.4 instead of %40): "))
    
    if(FR <= 0) : 
        raise ValueError("Fertility rate can't be smaller than or equal to zero due to simulation purposes. ")
        
    #######Number of generations#######
    NGEN = int(input("Enter number of  generations: "))
    
    if(NGEN < 1) : 
        raise ValueError("Number of generations can't be smaller than 1 due to simulation purposes.")
        
    print("")        
    simulation('One Child', N, MFR, FR, NGEN)
    simulation('One Son', N, MFR, FR, NGEN)
    
    simulation_status = int(input("If you want to do another simulation enter 1, enter 0 otherwise: "))
    
    if(simulation_status != 0 and simulation_status != 1) : 
        raise ValueError("Invalid input. You should have entered either a 0 or 1.")
    
    elif(simulation_status == 0) : 
        print("\n- - - - - - - - - - Simulation completed - - - - - - - - - -")
    

Enter initial population: 10000
Enter Male/Female ratio: 0.51
Enter fertility rate with as fraction value (e.g 0.4 instead of %40): 0.89
Enter number of  generations: 10

One Child  Scenario:
- - - - - - - - - -
Generation  1  Statistics: 
No. of MALES :  5100
No. of FEMALES:  4900
No. of couples :  4900
No. of Parents :  4361
No. of babies Total :  4361
No. of baby girls :  2160
No. of baby boys :  2201 

Generation  2  Statistics: 
No. of MALES :  2201
No. of FEMALES:  2160
No. of couples :  2160
No. of Parents :  1922
No. of babies Total :  1922
No. of baby girls :  979
No. of baby boys :  943 

Generation  3  Statistics: 
No. of MALES :  943
No. of FEMALES:  979
No. of couples :  943
No. of Parents :  839
No. of babies Total :  839
No. of baby girls :  414
No. of baby boys :  425 

Generation  4  Statistics: 
No. of MALES :  425
No. of FEMALES:  414
No. of couples :  414
No. of Parents :  368
No. of babies Total :  368
No. of baby girls :  182
No. of baby boys :  186 

Generation  

Generation  9  Statistics: 
No. of MALES :  95325
No. of FEMALES:  92722
No. of couples :  92722
No. of Parents :  82522
No. of babies Total :  153380
No. of baby girls :  75839
No. of baby boys :  77541 

Generation  10  Statistics: 
No. of MALES :  77541
No. of FEMALES:  75839
No. of couples :  75839
No. of Parents :  67496
No. of babies Total :  125449
No. of baby girls :  62158
No. of baby boys :  63291 

--------- SURE -  0:00:12.693011 

If you want to do another simulation enter 1, enter 0 otherwise: 0

- - - - - - - - - - Simulation completed - - - - - - - - - -


## RUNNING TIME ANALYSIS 

#### First input's (n1 = 10000) execution times were 0:00:00.007612 and 0:00:00.132685 . While, second input's (n2 = 1000000) were 0:00:00.158322 and 0:00:12.693011   Even though input size was multiplied with 100, running time of n2 is only 89,46 times of n1's running time. Therefore code works efficiently for larger input.

## SCENARIO ANALYSIS 

#### 1st run (n = 10000) : 
#### Final population (10th generation) : One Child -> 3 , One Son -> 2543

#### 2nd run (n = 1000000) : 
#### Final population (10th generation) : One Child -> 882, One Son -> 278829


#### One Child Policy's effect is more drastic than expected. If it were to be carried out for a long time, population would dramatically decrease. One Son policy seems much more stable and manageable. In China's actual case, over-representation of males in the population problem was caused by different reasons which won't be discussed here. However, it can be seen that one son policy also causes a slight difference between the male-female representation. Overall, one son policy seems more applicable.